File: src/com/reus/leetcode/ThreeSum.java
package com.reus.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * https://leetcode.com/problems/3sum/
 */
public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        if (nums.length == 0) {
            return new ArrayList<>();
        }
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i - 1 >= 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int sumToFind = 0 - nums[i];
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int curSum = nums[left] + nums[right];
                if (curSum < sumToFind) {
                    left++;
                } else if (curSum > sumToFind) {
                    right--;
                } else {
                    result.add(List.of(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                }
            }
        }
        return result;
    }
}

File: src/com/reus/leetcode/google/arrays/ThreeSum.kt
package com.reus.leetcode.google.arrays

/**
 * https://leetcode.com/problems/3sum/
 */
class ThreeSum {
    fun threeSum(nums: IntArray): List<List<Int>> {
        val sorted = nums.sorted()
        val result = mutableSetOf<List<Int>>()
        for (i in 0..sorted.size - 1) {
            val curEntry: List<Int> = mutableListOf(sorted[i])
            val rest = sorted.takeLast(sorted.size - i - 1).toIntArray()
            val sumToFind = 0 - sorted[i]
            val toAdd: Set<List<Int>> = twoSum(rest, sumToFind)
            if (toAdd.isEmpty().not()) {
                result.addAll(toAdd.map { it + curEntry })
            }
        }
        return result.toList()
    }

    private fun twoSum(nums: IntArray, toFind: Int): Set<List<Int>> {
        if (nums.size < 2) {
            return emptySet()
        }
        var left = 0
        var right = nums.size - 1
        val result = mutableSetOf<List<Int>>()
        while (left < right) {
            val curSum = nums[left] + nums[right]
            if (curSum == toFind) {
                result.add(listOf(nums[left], nums[right]))
                left++
            } else if (curSum > toFind) {
                right--
            } else {
                left++
            }
        }
        return result
    }
}

fun main() {
    ThreeSum().threeSum(intArrayOf(-1, 0, 1, 2, -1, -4))
}